ql/src/java/org/apache/hadoop/hive/ql/optimizer/SharedWorkOptimizer.java

```diff
@@ -248,6 +251,32 @@ public ParseContext transform(ParseContext pctx) throws SemanticException {
     return pctx;
   }
 
+  private static List<List<TableScanOperator>> groupTableScanOperators(List<Entry<String, Long>> sortedTables,
+      ArrayListMultimap<String, TableScanOperator> tableNameToOps, int batchSize) {
+    if (batchSize == -1) {
+      return Collections.singletonList(sortedTables.stream().map(Entry::getKey)
+          .flatMap(tableName -> tableNameToOps.get(tableName).stream()).collect(Collectors.toList()));
```

Comment:
This puts all TS ops into a single List, regardless of their source. Maybe you intend the following code?
```
    if (batchSize == -1) {
      return sortedTables.stream()
          .map(entry -> tableNameToOps.get(entry.getKey()))
          .collect(Collectors.toList());
    }
```